engine/table/src/main/java/io/deephaven/engine/table/impl/asofjoin/StaticAsOfJoinStateManagerTypedBase.java

```diff
@@ -0,0 +1,390 @@
+        final MutableInt hashOffset;
+
+        private StaticBuildHandler(final boolean isLeftSide, final LongArraySource hashSlots,
+                                   final MutableInt hashOffset) {
+            this.hashSlots = hashSlots;
+            this.isLeftSide = isLeftSide;
+            this.hashOffset = hashOffset;
+        }
+
+        @Override
+        public void doBuild(RowSequence chunkOk, Chunk<Values>[] sourceKeyChunks) {
+            hashSlots.ensureCapacity(hashOffset.intValue() + chunkOk.intSize());
+            if (isLeftSide) {
+                buildFromLeftSide(chunkOk, sourceKeyChunks, hashSlots, hashOffset);
+            } else {
+                buildFromRightSide(chunkOk, sourceKeyChunks, hashSlots, hashOffset);
+            }
+        }
+    }
+
+    @Override
+    public int buildFromLeftSide(RowSequence leftRowSet, ColumnSource<?>[] leftSources,
+            @NotNull final LongArraySource addedSlots) {
+        if (leftRowSet.isEmpty()) {
+            return 0;
+        }
+        try (final BuildContext bc = makeBuildContext(leftSources, leftRowSet.size())) {
+            final MutableInt slotCount = new MutableInt(0);
+            buildTable(bc, leftRowSet, leftSources, new StaticBuildHandler(true, addedSlots, slotCount));
+            return slotCount.intValue();
+        }
+    }
+
+    @Override
+    public int buildFromRightSide(RowSequence rightRowSet, ColumnSource<?>[] rightSources,
+            @NotNull final LongArraySource addedSlots) {
+        if (rightRowSet.isEmpty()) {
+            return 0;
+        }
+        try (final BuildContext bc = makeBuildContext(rightSources, rightRowSet.size())) {
+            final MutableInt slotCount = new MutableInt(0);
+            buildTable(bc, rightRowSet, rightSources, new StaticBuildHandler(false, addedSlots, slotCount));
+            return slotCount.intValue();
+        }
+    }
+
+    protected void probeTable(
+            final ProbeContext pc,
+            final RowSequence probeRows,
+            final boolean usePrev,
+            final ColumnSource<?>[] probeSources,
+            final TypedHasherUtil.ProbeHandler handler) {
+        try (final RowSequence.Iterator rsIt = probeRows.getRowSequenceIterator()) {
+            // noinspection unchecked
+            final Chunk<Values>[] sourceKeyChunks = new Chunk[probeSources.length];
+
+            while (rsIt.hasMore()) {
+                final RowSequence chunkOk = rsIt.getNextRowSequenceWithLength(pc.chunkSize);
+
+                if (usePrev) {
+                    getPrevKeyChunks(probeSources, pc.getContexts, sourceKeyChunks, chunkOk);
+                } else {
+                    getKeyChunks(probeSources, pc.getContexts, sourceKeyChunks, chunkOk);
+                }
+
+                handler.doProbe(chunkOk, sourceKeyChunks);
+
+                pc.resetSharedContexts();
+            }
+        }
+    }
+
+    private class StaticProbeHandler implements TypedHasherUtil.ProbeHandler {
```

Comment:
The two problems look different enough that you could have two classes pretty easily. One of the advantages of the javapoet approach is we can have as many probe/build functions as we want.  The other code had to rely on a single function with a bunch of flags, because we were just copying the file and replacing regions.  We can be cleaner in this version.

For the build, the parameters are identical so I can see why you woudl want to use the isLeft; but here I think we should change it.